### PR TITLE
Prerequisites for supporting domain detection classes

### PR DESCRIPTION
**Pull request by robertbartel**
A few prereqs to keep the commits cleaner.  Notably:

- Extending _Reader_ to support repeatable reads
- Making _ResultIndicator_ able to behave as a bool
- Fixing and improving geopackage hydrofabric implementation and test data
- Adding some data formats

### CODE REVIEW

File: python/lib/core/dmod/core/meta_data.py
Comment (aaraney) [inline python/lib/core/dmod/core/meta_data.py:202]: How do these interact with the `Dataset`'s `DataCategory`? At least, `EMPTY` and `GENERIC` seem related to that concept.

**Comment by aaraney** [comment]
Curious if the failures are related to removing Fiona as a dependency? I'm AFK, so I've not pulled the branch and verified locally.

File: python/lib/core/dmod/core/common/reader.py
Comment (robertbartel) [reply python/lib/core/dmod/core/common/reader.py:10]: Would the convention impose on us to also implement [`seekable()`](https://docs.python.org/3/library/io.html#io.IOBase.seek) (and, from that, `tell()` and `truncate()`)?

Comment (aaraney) [reply python/lib/core/dmod/core/common/reader.py:10]: Generally, if `seek` is an available method, it follows that the methods you listed are also available. But, I don't think that is a strict requirement. I try to keep `Protocol` interfaces as thin and narrowly scoped as necessary. So, in this case I would not include those methods.

**Review by aaraney — CHANGES_REQUESTED** [review]

File: python/lib/core/dmod/core/meta_data.py
Comment (robertbartel) [reply python/lib/core/dmod/core/meta_data.py:202]: Hmm ... I don't think there's any coupling, strictly speaking, between categories and either formats or domains.  For this in particular, I think that's fine:  it's reasonable to have a dataset in any category for which the data coverage is defined to be either empty or unknown. 

More generally, I can already see things with categories that need tweaking (the notion of an `OUTPUT` category is probably not valid, if we start to think of evaluation task needing modeling "output" as input, or forcing regridding jobs producing as "output" something belonging to `FORCING` category).  And we may not be using category sufficiently and/or consistently throughout our data modeling.  

Did you have something specific in mind when you noted this, or just a general concern?

File: python/lib/core/dmod/core/meta_data.py
Comment (robertbartel) [reply python/lib/core/dmod/core/meta_data.py:45]: I think we should keep it separate, at least for now (does any other data format use any versioning at the moment?).  

A lot of - maybe all - the other data DMOD works with is only fully valid if we assume its applied within some hydrofabric.  I suspect before too long we will want or need a hydrofabric version index in constraints defining the domain of regridded forcings or BMI init config datasets, to be able to tell if the `cat-1156` involved is actually the `cat-1156` we are interested in.  And that would be a more flexible way to constrain things than aligning the specific hydrofabric id.

File: python/lib/core/dmod/core/common/reader.py
Comment (robertbartel) [reply python/lib/core/dmod/core/common/reader.py:10]: I don't want to get too carried away with this, but I find the underlying design philosophy question genuinely interesting:  if the strength of adhering to convention is encoded expectations, (when) is it better avoid the convention if expectations aren't going to be fulfilled?

We can potentially continue the discussion, here or offline, but for the sake of practicality I'll plan to make the changes as you've suggested.

Comment (aaraney) [reply python/lib/core/dmod/core/common/reader.py:10]: Yeah! I think it's a really interesting question too! 

> if the strength of adhering to convention is encoded expectations, (when) is it better avoid the convention if expectations aren't going to be fulfilled?

I would argue that a function or method that supports a polymorphic argument should constrain the requirements of that argument to only what the function or method _needs_. So in this context, an interface (python Protocol) is simply a contract that some type `T` supports a given set of methods. It divorces the requirement of a `T` to be a certain type or subtype, but instead focuses only on the behaviors of the `T`. This enables constructing expectations (methods) for only what is _required_ and often reduces coupling. 

With this in mind, we can consider a convention as a _composition_ of expectations (methods) rather than a single set of expectations (traditional abc). So, the convention is not lost, we simply _only_ specify _our_ minimal set of requirements.

Comment (robertbartel) [reply python/lib/core/dmod/core/common/reader.py:10]: I largely agree, but a question follows from this on the strength of transitive expectations/requirements:  does `seek()` imply `seekable()`, or merely suggest it?  

As you point out, we could answer that by noting that the utilizing method doesn't strictly need to check whether the object is seek-able.  It _does_ need something that `seekable()` would provide (assurance that the data object can be read more than once), but not all of it, and what the method would need of it can be provided in a different way.   So, we don't really need the object to include `seekable()`.

The problem is that we could apply the same rationale with `seek()`.  The method does not need an object with full seek capabilities:  only an object that can do something that could be accomplished using `seek()`.  And we could provide that functionality to the object in a different, more focused way.  So we don't _really_ need `seek()` either. 

In other words, we would not be adopting that approach because it was the strictest possible constraint fulfillment, but rather because it's familiar:   this will probably let people (including us in the future) work with the object more quickly and easily.  But that fundamentally relies on expectations and assumptions for how things work.   And in that context, I'm not sure we can get a firm answer on the implication versus suggestion question.  

So the problem here to me is to approximating the balance point between convenience at the cost of risking reasonable-but-incorrect assumptions, versus avoiding ambiguity at the cost of familiarity.  I'm still not sure where that is with this.  Your option does feel more "Pythonic," though, so absent a more definitive mechanism, we can split the tie that way.

File: python/lib/core/dmod/core/meta_data.py
Comment (robertbartel) [reply python/lib/core/dmod/core/meta_data.py:202]: There definitely is conceptually, just not one that's explicitly defined somewhere.  

I haven't tested that scenario, but I expect it would allow the ``HYDROFABRIC`` category dataset to have a defined domain with the ``AORC_CSV`` format.   I'm pretty sure the same thing would happen if the user tried to provided the domain manually when creating a dataset, though.  So this is definitely a problem, but outside the scope of the automatic detection work.  I've opened #567 specifically for it.

**Comment by robertbartel** [comment]
I'm working to see what's going on with the tests.

File: python/lib/core/dmod/core/meta_data.py
Comment (robertbartel) [reply python/lib/core/dmod/core/meta_data.py:47]: I've opened #569 for tracking this in the future.

**Comment by robertbartel** [comment]
@aaraney, I think all the review comments have been properly addressed either in conversation, separate issues, or additional commits, and I have the tests passing now also, but let me know if I've missed anything.

File: python/lib/core/dmod/core/meta_data.py
Comment (aaraney) [reply python/lib/core/dmod/core/meta_data.py:47]: Thanks for opening something to track this!

**Review by aaraney — CHANGES_REQUESTED** [review]
I missed that `ReadSeeker` also needs to inherit `Protocol`. @christophertubbs, can you confirm this?

**Comment by aaraney** [comment]
Sorry, I didn't catch this until after i'd approved. I went to take a look at #565 and in passing noticed my mistake.

File: python/lib/core/dmod/core/common/reader.py
Comment (aaraney) [inline python/lib/core/dmod/core/common/reader.py:17]: Fairly certain this needs to be changed to the following. @christophertubbs, can you verify that? 

```suggestion
@runtime_checkable
class ReadSeeker(Reader, Seeker, Protocol):
```

Comment (christophertubbs) [reply python/lib/core/dmod/core/common/reader.py:17]: Yeah, the suggested replacement looks good. Want me to hit that commit button or should we wait on Bobby?

**Review by aaraney — APPROVED** [review]

**Pull request merged**
